File: sicp/1.7.py
def sqrt_iter(old_guess, guess, x):

    if (is_good_enough(old_guess, guess)):
        return guess
    return sqrt_iter(guess, improve_guess(guess, x), x)

def is_good_enough(old_guess, guess):
    return abs(old_guess - guess) < 0.001

def improve_guess(guess, x):
    return average(x / guess, guess)

def average(x, y):
    return x + y / 2

def better_sqrt_iter(n):
    return sqrt_iter(1, 0.5, n)

File: sicp_v2/1.7.rb
require "test/unit"

def average(x, y)
    return (x + y) / 2   
end


def is_good_enough(previous, guess)
    return ((previous - guess) / guess).abs < 0.001
end

def improve_sqrt(previous, guess, x)
    if is_good_enough previous, guess
        return guess
    end
    return improve_sqrt(guess, average(x / guess, x).to_f, x)
end

class TestSimpleNumber < Test::Unit::TestCase

    def test_equal
      assert_equal(1, improve_sqrt(0.5, 1, 2).floor)
      assert_equal(2, improve_sqrt(0.5, 1, 4).floor)
    end
end
